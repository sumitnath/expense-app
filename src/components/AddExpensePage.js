import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses'
import { startAddExpense } from '../actions/expenses';
// getting action geberator 
export class AddExpensePage extends React.Component{
  onSubmit=(expense)=>{
    //console.log(expense);
   // props.dispatch(addExpense(expense))
   this.props.startAddExpense(expense)
    this.props.history.push('/')
  };
  render(){
  return (
    <div>
    <h1> Add Expenses</h1>
    <ExpenseForm 
    onSubmit={this.onSubmit} />
  </div>
  )
 } 
}

  const mapDispatchToProps = (dispatch) =>{
    return {
      startAddExpense : (expense)=> dispatch(startAddExpense(expense))
        }
  }
//to pass the data to redux store we have to import connect 
// we donot need anything from the state we will keep 1st component of connect empty

export default connect(undefined, mapDispatchToProps)(AddExpensePage)