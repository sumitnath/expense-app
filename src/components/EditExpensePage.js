import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense} from '../actions/expenses'
import {removeExpense} from '../actions/expenses'

export class EditExpensePage extends React.Component {
  
 editExpense = (expense)=>{
      console.log('update',expense);
   //  this.props.dispatch(editExpense(this.props.expense.id,expense))
   this.props.increment(this.props.expense.id,expense)
     this.props.history.push('/')   
    }
  removeExpense =()=>{ 
    this.props.decrement({id:this.props.expense.id} );
    this.props.history.push('/')
  }
  render(){
return (
  <div>
  <h3> Edit Your expense</h3> 
    
    <ExpenseForm 
    expense ={this.props.expense}
    onSubmit={this.editExpense}
    />
    
  <button onClick ={this.removeExpense}>Remove</button>
    
  </div>
   )
  }
}
const mapDispatchToProps = (dispatch,props) => {
    return {
        increment: (id,expense) => dispatch(editExpense(id, expense)),
        decrement: (data) => dispatch(removeExpense(data))   
    }
  }

const mapStateToProps =(state,props ) =>{
  return{
  expense: state.expenses.find((expense)=>
   {return expense.id === props.match.params.id})
   }
  }
    
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)


