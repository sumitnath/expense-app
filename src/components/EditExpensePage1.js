import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses'
import { removeExpense } from '../actions/expenses'

// changing stateless function component to class based component in EditExpensePage 
const EditExpensePage = (props) => {
    console.log(props)
    return ( <div>
        <h3> Edit Your expense </h3>  
        <p> The id is: { props.match.params.id } </p>
         <ExpenseForm expense = { props.expense }
        onSubmit = {
            (expense) => {
                console.log('update', expense);
                props.dispatch(editExpense(this.props.expense.id,expense))
                props.history.push('/')
            }
        }
        />

        <button onClick = {() => {
                props.dispatch(removeExpense({ id: props.expense.id }));
             props.history.push('/')
            }
        } > Remove </button>

         </div>
    );
}


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => { return expense.id === props.match.params.id })
    }
}



export default connect(mapStateToProps)(EditExpensePage)