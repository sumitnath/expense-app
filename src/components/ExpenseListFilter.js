import React from "react";
import { connect } from "react-redux";
import{setTextFilter,sortByDate,sortByAmount, setStartDate,setEndDate} from '../actions/filters'
import { DateRangePicker } from "react-dates";

// converting the stateless component to class component
//const ExpenseListFilter = (props) => (
export class ExpenseListFilter extends React.Component {
   state ={
      calenderFocused: null
   }
   onDatesChange =({startDate:startDate,endDate:endDate})=>{
     this.props.setStartDate(startDate);
     this.props.setEndDate(endDate);
   }
   onFocusChange = (calenderFocused)=>{
     this.setState(()=>({calenderFocused:calenderFocused}))
   }
   onTextChange = (e)=>{
             this.props.setTextFilter(e.target.value)
          //console.log(e.target.value)
         }
   onSortChange =   (e)=>{
          if(e.target.value === 'date'){
             this.props.sortByDate()
          }else if(e.target.value === 'amount'){
             this.props.sortByAmount()
          }
         }    
   render(){
      return(
         <div>
         <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
         <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
         <option value ="date">Date</option>
         <option value="amount">Amount</option>
         </select>
         <DateRangePicker
           startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
          // startDateId={this.props.filters.startDate.toString()} // PropTypes.string.isRequired,
           endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
          // endDateId={this.props.filters.endDate.toString()}// PropTypes.string.isRequired,
           onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
           focusedInput={this.state.calenderFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
           onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
           numberOfMonths= {1}
           isOutsideRange= {()=>false}
           showClearDates={true} />
              </div>
      )
    }
   }

const mapStateToProps = (state)=>{
     return{
    filters:state.filters
 }
}
const mapDispatchToProps = (dispatch)=>{
   return {
      setEndDate: (endDate)=>dispatch(setEndDate(endDate)),
      setStartDate: (startDate)=>dispatch(setStartDate(startDate)),
      setTextFilter:(text)=>dispatch(setTextFilter(text)),
      sortByDate: ()=>dispatch(sortByDate()),
      sortByAmount:()=>dispatch(sortByAmount())
 
   }
}
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter)