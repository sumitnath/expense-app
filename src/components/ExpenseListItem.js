import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
//import { connect } from 'react-redux';
// we donot need the redux store as we have dont have remove here, so we remove connect()
// challenge remove remove button from  the ExpenseListItem.js and add it into EditExpensePage.js


const ExpenseListItem = ({ id, description, amount, createdAt }) => (

        <div>
        <Link to = { `/edit/${id}` } >
        <h3> { description } </h3>
         </Link> 
         <p> {numeral(amount/100).format('$0,0.00' )

         }
         
         -- {moment(createdAt).format('MMMMM Do, YYYY') } </p>


        </div>

    )
    //export default connect()(ExpenseListItem)
export default ExpenseListItem
