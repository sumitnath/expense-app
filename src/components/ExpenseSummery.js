import React from "react";
import { connect } from 'react-redux';
import numeral from "numeral";
import SelectExpenses from  '../selector/expenses';
import SelectExpenseTotal from  '../selector/expenses-total'


export const ExpenseSummery = ({expenseCount,expenseTotal}) =>{
const expenseWord = expenseCount === 1 ? 'expense' : 'expenses' ;
const formatedExpenseTotal = numeral(expenseTotal/100).format('$0.0.00')
 return (
    <div>
    <h1> Viewing {expenseCount} {expenseWord} totalling {formatedExpenseTotal}</h1>
    </div>
 );
};
const mapStateToProps = (state)=>{
    const visibleExpense = SelectExpenses(state.expenses, state.filters);
    return{
        expenseCount : visibleExpense.length,
        expenseTotal :SelectExpenseTotal(visibleExpense)
    }
};
export default connect(mapStateToProps)(ExpenseSummery);


