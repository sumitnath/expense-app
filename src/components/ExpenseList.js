import React from "react";
import { connect } from 'react-redux'
import ExpenseListItem from "./ExpenseListItem";
import SelectExpenses from '../selector/expenses'

export const ExpenseList = (props) => (
<div>
<h1>Expense List</h1>
{
    props.expenses.length === 0 ? (<p>No expenses</p>) :  (props.expenses.map((expense)=>{
        return  <ExpenseListItem key ={expense.id}{...expense}/>
     }))
}

</div>
)
const mapStateToProps = (state)=>{
    return {
        expenses: SelectExpenses(state.expenses,state.filters)
    }
}
export default  connect (mapStateToProps)(ExpenseList)
