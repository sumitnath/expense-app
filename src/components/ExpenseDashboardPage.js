import React from 'react';
//import { BrowserRouter,Route,Switch,Link,NavLink } from "react-router-dom";
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpenseSummery from './ExpenseSummery';

const ExpenseDashboardPage = () => (
  <div>
    This is from my dashboard component
    <ExpenseSummery/>
    <ExpenseListFilter/>
    <ExpenseList/>
  </div>
);


  export default ExpenseDashboardPage;