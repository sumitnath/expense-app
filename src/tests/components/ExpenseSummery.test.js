import React from "react";
import { shallow } from "enzyme";
import { ExpenseSummery } from "../../components/ExpenseSummery";

test('should correctly render Expense summary with 1 expense', () => {
            const wrapper = shallow( <ExpenseSummery expenseCount = { 1 }
                expenseTotal = { 233 }   />) 
                expect(wrapper).toMatchSnapshot();
            })

            
test('should correctly render Expense summary with multiple expense', () => {
            const wrapper = shallow( <ExpenseSummery expenseCount = { 21 }
                expenseTotal = { 225555533 }   />) 
                expect(wrapper).toMatchSnapshot();
            })