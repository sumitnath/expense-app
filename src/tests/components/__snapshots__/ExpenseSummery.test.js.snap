// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Expense summary with 1 expense 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseSummery
    expenseCount={1}
    expenseTotal={233}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h1>
         Viewing 
        1
         
        expense
         totalling 
        $2.3
</h1>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          " Viewing ",
          1,
          " ",
          "expense",
          " totalling ",
          "$2.3",
        ],
      },
      "ref": null,
      "rendered": Array [
        " Viewing ",
        1,
        " ",
        "expense",
        " totalling ",
        "$2.3",
      ],
      "type": "h1",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h1>
           Viewing 
          1
           
          expense
           totalling 
          $2.3
</h1>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            " Viewing ",
            1,
            " ",
            "expense",
            " totalling ",
            "$2.3",
          ],
        },
        "ref": null,
        "rendered": Array [
          " Viewing ",
          1,
          " ",
          "expense",
          " totalling ",
          "$2.3",
        ],
        "type": "h1",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should correctly render Expense summary with multiple expense 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseSummery
    expenseCount={21}
    expenseTotal={225555533}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h1>
         Viewing 
        21
         
        expenses
         totalling 
        $2255555.3
</h1>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          " Viewing ",
          21,
          " ",
          "expenses",
          " totalling ",
          "$2255555.3",
        ],
      },
      "ref": null,
      "rendered": Array [
        " Viewing ",
        21,
        " ",
        "expenses",
        " totalling ",
        "$2255555.3",
      ],
      "type": "h1",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h1>
           Viewing 
          21
           
          expenses
           totalling 
          $2255555.3
</h1>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            " Viewing ",
            21,
            " ",
            "expenses",
            " totalling ",
            "$2255555.3",
          ],
        },
        "ref": null,
        "rendered": Array [
          " Viewing ",
          21,
          " ",
          "expenses",
          " totalling ",
          "$2255555.3",
        ],
        "type": "h1",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
